File: code/abc/132e.py
def dijkstra(edge,start,end) :
    import heapq

    minCosts = [float('inf') for _ in range(len(edge))]

    nodeHeap = []
    heapq.heappush(nodeHeap, (0, start))  # (cost, node)

    while nodeHeap :
        cost, node = heapq.heappop(nodeHeap)

        if minCosts[node] <= cost :
            continue

        minCosts[node] = cost

        for nextNode , weight in edge[node] :
            newCost = cost + weight
            heapq.heappush(nodeHeap, (newCost, nextNode))
    dist = minCosts[endNode]
    if dist == float('inf') :
        return -1
    return dist

N , M = map(int,input().split())

# edges , costs
nodes = [[] for _ in range(3 * N)]

# 辺の入力
for _ in range(M) :
    fromNode , toNode = map(int,input().split())
    fromNode -= 1
    toNode -= 1
    fromNode *= 3
    toNode *= 3

    nodes[fromNode].append((toNode + 1,1))
    nodes[fromNode + 1].append((toNode + 2,1))
    nodes[fromNode + 2].append((toNode,1))

# スタート ゴール
startNode , endNode = map(int,input().split())
startNode -= 1
endNode -= 1
startNode *= 3
endNode *= 3

ans = dijkstra(nodes,startNode,endNode)

if ans == -1 :
    print(ans)
else :
    print(ans // 3)

File: code/abc/132e.cpp
#include <map>
#include <set>
#include <vector>
#include <algorithm>
#include <iostream>
#include <bitset>
#include <cassert>
#include <queue>
#include <random>
#include <stack>
#include <iomanip>

using namespace std;

typedef unsigned int uint;
typedef long long llong;
typedef unsigned long long ullong;
typedef long double ldouble;

typedef vector<llong> vecllong;
typedef vector<vecllong> vvecllong;

const llong MOD = 1e9 + 7;
const llong INF = 1e17;

#define FOR(i, n) for (llong i = 0; i < n; i++)
#define FORS(i, a, b) for (llong i = a; i < b; i++)
#define FORR(i, n) for (llong i = n; i > 0; i++)

llong dijkstra(llong start, llong end, vector<vector<pair<llong, llong>>> edge) {
    llong minCost[edge.size()];
    for (llong i = 0; i < edge.size(); i++) {
        minCost[i] = INF;
    }

    priority_queue<
        pair<llong, llong>,  // cost, index
        vector<pair<llong, llong>>,
        greater<pair<llong, llong>>
    > nodeHeap;

    nodeHeap.push(make_pair(0, start));

    while (!nodeHeap.empty()) {
        llong cost = nodeHeap.top().first;
        llong node = nodeHeap.top().second;
        nodeHeap.pop();

        if (minCost[node] <= cost) {
            continue;
        }

        minCost[node] = cost;

        for_each(edge[node].begin(), edge[node].end(), [&](pair<llong, llong> e){
            llong nextNode = e.first;
            llong newCost = cost + e.second;
            nodeHeap.push(make_pair(newCost, nextNode));
        });
    }

    return minCost[end];
}

int main(void) {
    llong N, M;
    cin >> N >> M;  // N頂点 M辺

    vector<vector<pair<llong, llong>>> edge(N*3);

    for (llong i = 0; i < M; i ++) {
        llong fromNode, toNode;
        llong cost = 1;
        cin >> fromNode >> toNode;
        fromNode--;
        toNode--;
        edge[fromNode*3].push_back(make_pair(toNode*3 + 1, cost));
        edge[fromNode*3 + 1].push_back(make_pair(toNode*3 + 2, cost));
        edge[fromNode*3 + 2].push_back(make_pair(toNode*3, cost));
    }

    llong start, end;
    cin >> start >> end;

    start--;
    end--;

    llong dist = dijkstra(start*3, end*3, edge);

    if (dist == INF) {
        cout << "-1" << endl;
    } else {
        cout << dist / 3 << endl;
    }

    return 0;
}

